### PR TITLE
Atomic arithmetic intrinsics

### PR DESCRIPTION
**Pull request by TheNumbat**
Adding builtins for the equivalent of GCC's `__sync_fetch_and_add/sub` and `__sync_bool_compare_and_swap`.

The pointer type may be native_pointer, ext_pointer, or bigstring with offset. The integer type may be untagged int, unboxed int64, unboxed int32, or unboxed nativeint.

The builtins are not translated on ARM (but it would now be easy to add ARM backend support - armv8 provides single instructions for each of these operations). 

The unboxed int64 builtins are not translated on 32 bit architectures.

### CODE REVIEW

File: backend/mach.mli [hunk 1]
```diff
@@ -27,6 +27,9 @@ type integer_comparison =
     Isigned of Cmm.integer_comparison
   | Iunsigned of Cmm.integer_comparison
 
+type atomic_integer_operation =
+  Ifetchadd | Ifetchsub | Ifetchand | Ifetchor | Ifetchxor | Icompareandswap
```
Comment (gretay-js) [inline backend/mach.mli:31]: It would be better to use  the same names as in `Cmm.atomic_op`.

File: backend/cmm.mli [hunk 2]
```diff
@@ -104,6 +104,8 @@ type rec_flag = Nonrecursive | Recursive
 
 type prefetch_temporal_locality_hint = Nonlocal | Low | Moderate | High
 
+type atomic_op = Fetch_add | Fetch_sub | Fetch_and | Fetch_or | Fetch_xor | CAS
```
Comment (gretay-js) [inline backend/cmm.mli:107]: I think it would be easier to maintain if the names here match the names of intrinsics in cmm_helpers, so I'd prefer `Fetch_and_add` and `Compare_and_swap`.

File: backend/selectgen.ml [hunk 3]
```diff
@@ -642,6 +644,12 @@ method select_operation op args _dbg =
   | (Cintoffloat, _) -> (Iintoffloat, args)
   | (Cvalueofint, _) -> (Ivalueofint, args)
   | (Cintofvalue, _) -> (Iintofvalue, args)
+  | (Catomic {op; size}, [src; dst]) ->
```
Comment (gretay-js) [inline backend/selectgen.ml:647]: Something like this that explicitly lists the cases would be safer:
```suggestion
  | (Catomic {op=(Fetch_add|Fetch_sub) as op; size}, [src; dst]) ->
```

File: backend/amd64/emit.mlp [hunk 4]
```diff
@@ -698,6 +698,61 @@ let emit_push_trap_label handler =
...
+      true, arg instr 1, addressing addr QWORD instr 2
+    | Thirtytwo_signed | Thirtytwo_unsigned ->
+      false, arg32 instr 1, addressing addr DWORD instr 2
+    | Byte_unsigned | Byte_signed
+    | Sixteen_unsigned | Sixteen_signed
+    | Single | Double -> Misc.fatal_error "Atomic cas got non 32/64 bit size."
+  in
+  match qword with
+  | true -> I.lock_cmpxchgq set_to dst; I.set E res
+  | false -> I.lock_cmpxchgl set_to dst; I.set E res
```
Comment (gretay-js) [inline backend/amd64/emit.mlp:716]: The result needs to be zero extended.

File: backend/cmm_helpers.ml [hunk 5]
```diff
@@ -3313,6 +3319,41 @@ let prefetch_offset ~is_write locality (arg1, arg2) dbg =
...
 
+let native_pointer_cas size (dst, compare_with, set_to) dbg =
+  let op = Catomic { op = CAS; size } in
+  if_operation_supported op ~f:(fun () ->
+      Cop (op, [compare_with; set_to; dst], dbg))
+
+let ext_pointer_cas size (dst, compare_with, set_to) dbg =
+  let op = Catomic { op = CAS; size } in
+  if_operation_supported op ~f:(fun () ->
+      Cop (op, [compare_with; set_to; int_as_pointer dst dbg], dbg))
```
Comment (gretay-js) [inline backend/cmm_helpers.ml:3330]: Is this instruction setting the pointer itself or the memory block pointed by it? If latter, I think we need a load here (similarly to fetch_and_add), and I'm not sure how the tests pass.

**Review by gretay-js — CHANGES_REQUESTED** [review]

File: backend/amd64/selection.ml [hunk 6]
```diff
@@ -94,8 +94,23 @@ let rdx = phys_reg 4
...
+      let arg = Array.copy arg in
+      arg.(0) <- rax;
+      (arg, res)
+  | Iintop_atomic {op = Ifetchadd|Ifetchsub; size = _; addr = _} ->
+      (* first arg must be the same as res.(0) *)
+      let arg = Array.copy arg in
+      arg.(0) <- res.(0);
+      (arg, res)
+  | Iintop_atomic {op = Ifetchand|Ifetchor|Ifetchxor; size = _; addr = _} ->
+      (* first arg and result must be rax, and it uses rcx/rdx *)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/selection.ml:110]: Removed the bitwise ops

File: backend/amd64/emit.mlp [hunk 7]
```diff
@@ -698,6 +698,61 @@ let emit_push_trap_label handler =
...
+      false, arg32 instr 0, addressing addr DWORD instr 1
+    | Byte_unsigned | Byte_signed
+    | Sixteen_unsigned | Sixteen_signed
+    | Single | Double -> Misc.fatal_error "Atomic got non 32/64 bit size."
+  in
+  match qword, op with
+  | true, Ifetchadd -> I.lock_xaddq src dst
+  | false, Ifetchadd -> I.lock_xaddl src dst
+  | true, Ifetchsub -> I.neg src; I.lock_xaddq src dst
+  | false, Ifetchsub -> I.neg src; I.lock_xaddl src dst
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:752]: Removed fetch_and_sub - cmmgen now transforms it to add

File: backend/amd64/emit.mlp [hunk 8]
```diff
@@ -698,6 +698,61 @@ let emit_push_trap_label handler =
...
+      false, arg32 instr 1, addressing addr DWORD instr 2
+    | Byte_unsigned | Byte_signed
+    | Sixteen_unsigned | Sixteen_signed
+    | Single | Double -> Misc.fatal_error "Atomic cas got non 32/64 bit size."
+  in
+  match qword with
+  | true -> I.lock_cmpxchgq set_to dst; I.set E res
+  | false -> I.lock_cmpxchgl set_to dst; I.set E res
+
+let emit_bit_atomic op qword dst =
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:718]: Removed

File: backend/mach.mli (hunk 1 above)
Comment (TheNumbat) [reply backend/mach.mli:31]: Changed

File: backend/cmm.mli (hunk 2 above)
Comment (TheNumbat) [reply backend/cmm.mli:107]: Changed

File: backend/cmm_helpers.ml (hunk 5 above)
Comment (TheNumbat) [reply backend/cmm_helpers.ml:3330]: Hmm, not sure what you mean - it's setting the memory pointed to by the ext pointer. It works in the same manner as the native_pointer case - it just needs to decode the ext pointer first (int_as_pointer).

File: backend/amd64/emit.mlp [hunk 9]
```diff
@@ -698,6 +698,61 @@ let emit_push_trap_label handler =
...
+  (match op with
+  | Ifetchand -> I.and_ rdx rcx; (* input|rdx & rcx -> rcx *)
+  | Ifetchor -> I.or_ rdx rcx;   (* input|rdx | rcx -> rcx *)
+  | Ifetchxor -> I.xor rdx rcx;  (* input|rdx ^ rcx -> rcx *)
+  | _ -> Misc.fatal_error "Atomic bitwise op got non bitwise op.");
+  cmpxchg rcx dst;  (* rcx -> (dst) if (dst) = rax *)
+  I.jne (label lbl) (* else restart *)
+
+let emit_atomic instr op size addr =
+  let qword, src, dst =
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:738]: Refactored

File: backend/amd64/emit.mlp [hunk 10]
```diff
@@ -698,6 +698,61 @@ let emit_push_trap_label handler =
   traps.push_traps <- lbl::traps.push_traps;
   traps.enter_traps <- Int.Set.add handler traps.enter_traps
 
+let emit_cas instr size addr =
+  (* compare_with is in rax *)
+  let res = res8 instr 0 in
+  let qword, set_to, dst =
+    match size with
+    | Word_int | Word_val ->
+      true, arg instr 1, addressing addr QWORD instr 2
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:707]: Refactored

File: backend/amd64/emit.mlp [hunk 11]
```diff
@@ -698,6 +698,61 @@ let emit_push_trap_label handler =
...
+      true, arg instr 0, addressing addr QWORD instr 1
+    | Thirtytwo_signed | Thirtytwo_unsigned ->
+      false, arg32 instr 0, addressing addr DWORD instr 1
+    | Byte_unsigned | Byte_signed
+    | Sixteen_unsigned | Sixteen_signed
+    | Single | Double -> Misc.fatal_error "Atomic got non 32/64 bit size."
+  in
+  match qword, op with
+  | true, Ifetchadd -> I.lock_xaddq src dst
+  | false, Ifetchadd -> I.lock_xaddl src dst
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:750]: Combined

File: backend/selectgen.ml (hunk 3 above)
Comment (TheNumbat) [reply backend/selectgen.ml:647]: Listed

File: backend/mach.ml [hunk 12]
```diff
@@ -67,6 +70,8 @@ type operation =
                 mode : Lambda.alloc_mode }
   | Iintop of integer_operation
   | Iintop_imm of integer_operation * int
+  | Iintop_atomic of { op : atomic_integer_operation;
```
Comment (gretay-js) [inline backend/mach.ml:73]: We can use `Cmm.atomic_integer_operation` here.

File: backend/cmm_helpers.ml [hunk 13]
```diff
@@ -3313,6 +3321,71 @@ let prefetch_offset ~is_write locality (arg1, arg2) dbg =
...
+        (fun src ->
+          bind "idx" arg2 (fun idx ->
+              bind "bs" arg1 (fun bs ->
+                  bind "bs_data"
+                    (Cop
+                       (Cload (Word_int, Mutable), [field_address bs 1 dbg], dbg))
+                    (fun bs_data ->
+                      bind "dst" (add_int bs_data idx dbg) (fun dst ->
+                          Cop (op, [src; dst], dbg)))))))
+
```
Comment (gretay-js) [inline backend/cmm_helpers.ml:3388]: Can you try to factor out some of the duplication into helper functions, hopefully voiding  polymorphic variants?

File: backend/selectgen.ml [hunk 14]
```diff
@@ -174,6 +174,7 @@ let oper_result_type = function
   | Calloc _ -> typ_val
   | Cstore (_c, _) -> typ_void
   | Cprefetch _ -> typ_void
+  | Catomic _ -> typ_int
```
Comment (gretay-js) [inline backend/selectgen.ml:177]: `Ifetch_and_add` is `typ_int` but `Icompare_and_swap` should be `typ_val` now that it returns a tagged bool.

File: backend/printcmm.ml [hunk 15]
```diff
@@ -101,6 +101,10 @@ let temporal_locality = function
   | Moderate -> "moderate"
   | High -> "high"
 
+let atomic_op = function
+  | Fetch_and_add -> "add"
+  | Compare_and_swap -> "compareandswap"
```
Comment (gretay-js) [inline backend/printcmm.ml:106]:
```suggestion
  | Fetch_and_add -> "fetch_and_add"
  | Compare_and_swap -> "compare_and_swap"
```

File: backend/cmm_helpers.ml (hunk 5 above)
Comment (gretay-js) [reply backend/cmm_helpers.ml]: ah, yes, it is fine, selectgen does the right thing here.

File: backend/mach.ml (hunk 12 above)
Comment (TheNumbat) [reply backend/mach.ml:73]: Changed

File: backend/cmm_helpers.ml (hunk 13 above)
Comment (TheNumbat) [reply backend/cmm_helpers.ml:3388]: Done

File: backend/x86_gas.ml [hunk 16]
```diff
@@ -105,7 +105,8 @@ let i0 b s = bprintf b "\t%s" s
 let i1 b s x = bprintf b "\t%s\t%a" s arg x
 let i1_s b s x = bprintf b "\t%s%s\t%a" s (suf x) arg x
 let i2 b s x y = bprintf b "\t%s\t%a, %a" s arg x arg y
-let i2_s b s x y = bprintf b "\t%s%s\t%a, %a" s (suf y) arg x arg y
+let i2_sx b s x y = bprintf b "\t%s%s\t%a, %a" s (suf x) arg x arg y
```
Comment (gretay-js) [inline backend/x86_gas.ml:108]: I think you should keep pre-existing functions with the same (bad) name to avoid unrelated changes in this PR.

**Review by gretay-js — APPROVED** [review]
Looks good to me. 

@xclerc  this is ready for you to review.

File: backend/x86_gas.ml (hunk 16 above)
Comment (TheNumbat) [reply backend/x86_gas.ml:109]: :+1:

**Review by xclerc — COMMENTED** [review]
I did a fairly superficial review; most
comments between parentheses can
be safely ignored. As always, it is a bit
annoying to have no tests.

File: backend/arm64/emit.mlp [hunk 17]
```diff
@@ -493,6 +493,9 @@ module BR = Branch_relaxation.Make (struct
       num_instructions_for_intconst n
     | Lop (Iconst_float _) -> 2
     | Lop (Iconst_symbol _) -> 2
+    | Lop (Iintop_atomic _) ->
+      (* Never generated; builtins are not yet translated to atomics *)
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/arm64/emit.mlp:497]: there are arm64 instructions for it.

File: backend/x86_binary_emitter.ml [hunk 18]
```diff
@@ -1220,6 +1220,34 @@ let emit_LEA b dst src =
...
+  | ((Mem _ | Mem64_RIP _) as rm), Reg64 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b rexw [ 0x0F; 0xB1 ] rm reg
+  | ((Mem _ | Mem64_RIP _) as rm), Reg32 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b no_rex [ 0x0F; 0xB1 ] rm reg
+  | _ ->
+      Format.eprintf "lock cmpxchg src=%a dst=%a@." print_old_arg src print_old_arg dst;
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/x86_binary_emitter.ml:1234]: :+1:

File: backend/x86_binary_emitter.ml [hunk 19]
```diff
@@ -1220,6 +1220,34 @@ let emit_LEA b dst src =
...
+  | ((Mem _ | Mem64_RIP _) as rm), Reg64 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b rexw [ 0x0F; 0xC1 ] rm reg
+  | ((Mem _ | Mem64_RIP _) as rm), Reg32 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b no_rex [ 0x0F; 0xC1 ] rm reg
+  | _ ->
+      Format.eprintf "lock xadd src=%a dst=%a@." print_old_arg src print_old_arg dst;
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/x86_binary_emitter.ml:1248]: :+1:

File: backend/x86_binary_emitter.ml [hunk 20]
```diff
@@ -1220,6 +1220,34 @@ let emit_LEA b dst src =
...
+let emit_lock_cmpxchg b dst src =
+  match (dst, src) with
+  | ((Mem _ | Mem64_RIP _) as rm), Reg64 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b rexw [ 0x0F; 0xB1 ] rm reg
+  | ((Mem _ | Mem64_RIP _) as rm), Reg32 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b no_rex [ 0x0F; 0xB1 ] rm reg
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/x86_binary_emitter.ml:1232]: :+1:

File: backend/x86_binary_emitter.ml [hunk 21]
```diff
@@ -1220,6 +1220,34 @@ let emit_LEA b dst src =
...
+let emit_lock_xadd b dst src =
+  match (dst, src) with
+  | ((Mem _ | Mem64_RIP _) as rm), Reg64 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b rexw [ 0x0F; 0xC1 ] rm reg
+  | ((Mem _ | Mem64_RIP _) as rm), Reg32 reg ->
+      let reg = rd_of_reg64 reg in
+      buf_int8 b 0xF0;
+      emit_mod_rm_reg b no_rex [ 0x0F; 0xC1 ] rm reg
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/x86_binary_emitter.ml:1246]: :+1:

File: backend/printcmm.ml [hunk 22]
```diff
@@ -216,6 +220,7 @@ let operation d = function
   | Cprefetch { is_write; locality; } ->
     Printf.sprintf "prefetch is_write=%b prefetch_temporal_locality_hint=%s"
       is_write (temporal_locality locality)
+  | Catomic { op } -> Printf.sprintf "atomic %s" (atomic_op op)
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/printcmm.ml:223]: :+1:

File: backend/printmach.ml [hunk 23]
```diff
@@ -179,6 +179,16 @@ let operation' ?(print_reg = reg) op arg ppf res =
         fprintf ppf "%a%s%a" reg arg.(0) (intop op) reg arg.(1)
       end
   | Iintop_imm(op, n) -> fprintf ppf "%a%s%i" reg arg.(0) (intop op) n
+  | Iintop_atomic {op = Compare_and_swap; size; addr} ->
+    fprintf ppf "lock cas %s[%a] ?%a %a"
+      (Printcmm.chunk size)
+      (Arch.print_addressing reg addr) (Array.sub arg 2 (Array.length arg - 1))
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/printmach.ml:185]: oh, indeed...

File: backend/arm64/emit.mlp [hunk 24]
```diff
@@ -493,6 +493,9 @@ module BR = Branch_relaxation.Make (struct
...
+      assert false
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/arm64/emit.mlp:498]: Yes - the only place the instruction can be generated is from the atomic builtins, which are skipped on arm

File: backend/cmm.mli [hunk 25]
```diff
@@ -196,6 +198,7 @@ and operation =
   | Cctz of { arg_is_non_zero: bool; }
   | Cpopcnt
   | Cprefetch of { is_write: bool; locality: prefetch_temporal_locality_hint; }
+  | Catomic of { op: atomic_op; size : memory_chunk }
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cmm.mli:201]: :+1:

File: backend/cfg/cfg_intf.ml [hunk 26]
```diff
@@ -66,6 +66,11 @@ module S = struct
     | Store of Cmm.memory_chunk * Arch.addressing_mode * bool
     | Intop of Mach.integer_operation
     | Intop_imm of Mach.integer_operation * int
+    | Intop_atomic of
+        { op : Cmm.atomic_op;
+          size : Cmm.memory_chunk;
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/cfg/cfg_intf.ml:71]: :+1:

File: backend/mach.ml [hunk 27]
```diff
@@ -67,6 +67,8 @@ type operation =
                 mode : Lambda.alloc_mode }
   | Iintop of integer_operation
   | Iintop_imm of integer_operation * int
+  | Iintop_atomic of { op : Cmm.atomic_op; size : Cmm.memory_chunk;
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/mach.ml:70]: :+1:

File: backend/amd64/emit.mlp [hunk 28]
```diff
@@ -698,6 +698,29 @@ let emit_push_trap_label handler =
...
   traps.enter_traps <- Int.Set.add handler traps.enter_traps
 
+let emit_atomic instr op size addr =
+  let qword =
+    match size with
+    | Word_int | Word_val -> true
+    | Thirtytwo_signed | Thirtytwo_unsigned -> false
+    | Byte_unsigned | Byte_signed
+    | Sixteen_unsigned | Sixteen_signed
+    | Single | Double -> Misc.fatal_error "Atomic got non 32/64 bit size."
```
[the remark this answers is not in the document]
Comment (TheNumbat) [reply backend/amd64/emit.mlp:708]: :+1:
